### PR TITLE
api/news: get authors from algolia in primary

### PR DESCRIPTION
**Pull request by babygoat**
This patch adds the primary data source of authors(algolia) and
fallbacks to the mongo query if the index query results in error.

### CODE REVIEW

**Comment by babygoat** [comment]
Updated. Ready for review. Besides, the index data are uploaded to `developer-for-testing` application which can be utilized to integration testing.

File: controllers/news_v2.go [hunk 1]
```diff
@@ -358,18 +359,22 @@ func (nc *newsV2Controller) GetAuthors(c *gin.Context) {
 	q := news.ParseAuthorListQuery(c)
 
 	//TODO(babygoat): fetch from algolia first
```
Comment (taylrj) [inline controllers/news_v2.go]: Remove TODO comment ?

File: internal/news/author.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/algolia/algoliasearch-client-go/v3/algolia/opt"
+	"github.com/pkg/errors"
+)
+
+func GetAuthorWithIndex(index AlgoliaSearcher, q *Query) ([]Author, int, error) {
+	var authors []Author
+	res, err := index.Search(q.Filter.Name, opt.Offset(q.Offset), opt.Length(q.Limit))
+	if err != nil {
+		// fallback
+		return nil, -1, errors.WithStack(err)
```
Comment (taylrj) [inline internal/news/author.go:16]: Q: Why does it returns `-1` instead of `0` as the `nbHits` return value ?

File: storage/news_v2.go [hunk 3]
```diff
@@ -316,33 +316,52 @@ func (m *mongoStorage) getAuthors(ctx context.Context, stages []bson.D) <-chan f
...
+	case <-ctx.Done():
+		return 0, errors.WithStack(ctx.Err())
+	case res, ok := <-result:
+		switch {
+		case !ok:
+			return 0, errors.WithStack(ctx.Err())
+		case res.Error != nil:
+			return 0, res.Error
+		}
+		count = res.Content.(int64)
```
Comment (taylrj) [inline storage/news_v2.go:363]: Do we need to check whether the assertion is ok here ?

e.g.

```go
  if count, ok := res.Content.(int64); ok {
     return count, nil
  }
```

**Comment by babygoat** [comment]
updated.

File: storage/news_v2.go (hunk 3 above)
Comment (babygoat) [reply storage/news_v2.go:363]: We ensure the `res.Content` is always int64 by single sender(Ln350).

File: internal/news/author.go (hunk 2 above)
Comment (babygoat) [reply internal/news/author.go]: updated.

Comment (babygoat) [reply internal/news/author.go]: BTW, `-1` is used to mark the value is valid whenever the `err != nil`.

**Comment by babygoat** [comment]
Merge first for demo.

**Comment by taylrj** [comment]
LGTM

**Pull request merged**